backends/dpdk/spec.cpp

```diff
@@ -0,0 +1,550 @@
+}
+
+std::ostream &IR::DpdkHeaderType::toSpec(std::ostream &out) const {
+  out << "struct " << name << " {" << std::endl;
+  if (fields.empty()) {
+    out << "}" << std::endl;
+    return out;
+  }
+  for (auto it = fields.begin(); it != fields.end(); ++it) {
+    if (auto t = (*it)->type->to<IR::Type_Bits>())
+      out << "\tbit<" << t->width_bits() << ">";
+    else if (auto t = (*it)->type->to<IR::Type_Name>())
+      out << "\t" << t->path->name;
+    else if ((*it)->type->to<IR::Type_Boolean>())
+      out << "\tbool";
+    else {
+      std::cout << (*it)->type->node_type_name() << std::endl;
+      BUG("Unsupported type");
+    }
+    out << " " << (*it)->name;
+    out << std::endl;
+  }
+  out << "}" << std::endl;
+  return out;
+}
+
+std::ostream &IR::DpdkStructType::toSpec(std::ostream &out) const {
+  if (getAnnotations()->getSingle("__packet_data__")) {
+    for (auto it = fields.begin(); it != fields.end(); ++it) {
+      if (auto t = (*it)->type->to<IR::Type_Name>())
+        out << "header " << (*it)->name << " instanceof " << t->path->name;
+      else {
+        std::cout << (*it)->type->node_type_name() << std::endl;
+        BUG("Unsupported type");
+      }
+      out << std::endl;
+    }
+  } else {
+    out << "struct " << name << " {" << std::endl;
+    if (fields.empty()) {
+      out << "}" << std::endl;
+      return out;
+    }
+    for (auto it = fields.begin(); it != fields.end(); ++it) {
+      if (auto t = (*it)->type->to<IR::Type_Bits>())
+        out << "\tbit<" << t->width_bits() << ">";
+      else if (auto t = (*it)->type->to<IR::Type_Name>()) {
+        if (t->path->name == "error") {
+          out << "\tbit<8>";
+        } else {
+          out << "\t" << t->path;
+        }
+      } else if ((*it)->type->to<IR::Type_Error>()) {
+        // DPDK implements error type as bit<8>
+        out << "\tbit<8>";
+      } else if ((*it)->type->to<IR::Type_Boolean>()) {
+        // DPDK implements bool as bit<8>
+        out << "\tbit<8>";
+      } else {
+        std::cout << (*it)->type->node_type_name() << std::endl;
+        BUG("Unsupported type");
+      }
+      out << " " << (*it)->name;
+      out << std::endl;
+    }
+    out << "}" << std::endl;
+    if (getAnnotations()->getSingle("__metadata__")) {
+      out << "metadata instanceof " << name << std::endl;
+    }
+  }
+  return out;
+}
+
+std::ostream &IR::DpdkListStatement::toSpec(std::ostream &out) const {
+  out << "apply {" << std::endl;
+  out << "\trx m.psa_ingress_input_metadata_ingress_port" << std::endl;
+  for (auto s : statements) {
+    out << "\t";
+    s->toSpec(out) << std::endl;
+  }
+  out << "\ttx m.psa_ingress_output_metadata_egress_port" << std::endl;
+  out << "}" << std::endl;
+  return out;
+}
+
+std::ostream &IR::DpdkMovStatement::toSpec(std::ostream &out) const {
+  out << "mov " << DPDK::toStr(left) << " " << DPDK::toStr(right);
+  return out;
+}
+
+std::ostream &IR::DpdkAddStatement::toSpec(std::ostream &out) const {
```

Comment:
this is very strange, this does not use src1 at all.
If src1 must be identical to dst, then why store it at all?

Reply:
@gordonwucn please take a look.

Reply:
@hanw Since dpdk's binary operators only take in two arguments, DpdkBinaryStatement should only have two members _dst_ and _src_. And for every class that inherits DpdkBinaryStatement, the constructor function should also only take in two parameters _dst_ and _src_. Also, for every instantiation of these classes, two parameters should be passed in instead of three(src1 and src2's value should be identical for every class instantiation, so pick either of them as the value for src)